src/gsw_internal_funcs/mod.rs

```diff
@@ -1052,8 +1054,229 @@ mod test_gibbs {
+        );
+
+        let ans = g0p + TT * g.re;
+        return Ok(ans);
+
+    } else if nt == 1 && np == 1 {
+        
+        let tau_t2 = tau / T2;
+
+        let r2p = REC_PT * (R21 + 2.0 * R22 * dzi);
+
+        let g = r2p * (
+            ((1.0 + tau_t2)/(1.0 - tau_t2)).ln() - 2.0 * tau_t2
+        );
+
+        let ans = g.re;
+        return Ok(ans);
+
+    } else if nt == 2 && np == 0 { // start to rewrite without num-complex
+
+        let r2 = R20_GIBBS_ICE + dzi*(R21 + R22*dzi);
+        let r2_re: f64 = R20_GIBBS_ICE.re + dzi *(R21.re + R22.re*dzi);
+        let r2_im: f64 = R20_GIBBS_ICE.im + dzi * (R21.im + R22.im*dzi);
+
+        let g = R1_GIBBS_ICE * (
+            1.0/(T1 - tau) + 1.0/(T1 + tau) - 2.0/T1
+        ) + r2 * (
+            1.0/(T2 - tau)
+        );
+
+        // define helper variables
+        let a = T1.re - tau.re;
+        let b = T1.im - tau.im;
+        let c = T1.re + tau.re;
+        let d = T1.im + tau.im;
+        let e = T1.re;
+        let f = T1.im;
+        let h = T2.re - tau.re;
+        let j = T2.im - tau.im;
+
+        let g_re = (R1_GIBBS_ICE * (
+            1.0/(T1 - tau) + 1.0/(T1 + tau) - 2.0/T1
+        )).re + r2.re * (T2.re + tau.re);
+
+        // let g_re: f64 = R1_GIBBS_ICE.re * (
+        //     a/(a.powi(2) + b.powi(2)) + c/(c.powi(2) + d.powi(2)) - 2.0 * e/(e.powi(2) + f.powi(2))
+        // ) - R1_GIBBS_ICE.im * (
+        //     -b/(a.powi(2) + b.powi(2)) -d/(c.powi(2) + d.powi(2)) -2.0*f/(e.powi(2) + f.powi(2))
+        // )
+        // + (r2.re * h + r2.im * j) / (h.powi(2) + j.powi(2));
+
+        let ans = REC_TT * g_re;
+        return Ok(ans);
+
+    } else if nt == 0 && np == 2 {
+        
+        let sqrec_pt = REC_PT * REC_PT;
+
+        let tau_t2 = tau / T2;
+        let sqtau_t2 = tau_t2 * tau_t2;
+
+        let g0pp = sqrec_pt * (2.0*G02 + dzi*(6.0*G03 + 12.0*G04*dzi));
+
+        let r2pp = 2.0*R22*sqrec_pt;
+
+        let g = r2pp * (
+            tau * ((1.0 + tau_t2)/(1.0 - tau_t2)).ln()
+            + T2 * ((1.0 - sqtau_t2).ln() - sqtau_t2)
+        );
+
+        let ans = g0pp + TT * g.re;
+        return Ok(ans);
+    } else {
+        return Err(Error::OutOfBounds);
+    }
+}
+
+#[cfg(test)]
+mod test_gibbs_ice {
+
+    use num_complex::ComplexFloat;
+
+    use crate::gsw_internal_funcs::{gibbs_ice};
+    use crate::Error;
+
+
+    #[test]
+    fn exploration() {
+        // answers have not been checked for correctness, but are outputs of the function using num-complex
+        // (nt, np, ans)
+        let answers = [
+            (0, 0, 98.267598402919248),
+            (1, 0 , 1220.7886612999528),
+            (0, 1, 0.0010908434429264351),
+            (1, 1, 1.7436082496084962E-7),
+            (2, 0, -9.0784284079463049),
+            (0, 2, -1.2848482463976177E-13)
+        ];
+        for (nt, np, ans) in answers.iter() {
+            assert!((gibbs_ice(*nt, *np, 0.0, 0.0).unwrap() - *ans).abs() < f64::EPSILON);
```

Comment:
Note that [GA is failing in this test](https://github.com/castelao/GSW-rs/actions/runs/5282585889/jobs/9558811286?pr=51#step:4:339), but it is not clear which case it is failing. To help us, it would be best to add some information here.

Look at this example on [how to add a note](https://doc.rust-lang.org/std/macro.assert.html#examples) in the assert check. Probably add the coefficient that it is running (nt, np, and ans)

Reply:
Those notes will be super helpful! The failing test is nt = 2 and np = 0 since I tried removing num-complex from a section of it. Complex division has been most tricky, but I'll see how far I get by tomorrow.